binding: typechecks

muzarski opened the pull request:
Fix: https://github.com/scylladb/cpp-rust-driver/issues/142

## Motivation

Right now, cpp-rust-driver does not perform any typechecks during binding. This PR changes that and implements a typecheck logic based on cpp-driver's logic.

### Why not just use a builtin rust-driver's typechecks mechanism?
The reason for this is that cpp-driver has a bit different language (cpp/rust) to CQL type mapping. A simple example is an int64_t/i64 value which in cpp-driver can be mapped to 4 different CQL types.

Other reason is that cpp-driver allows users to define untyped tuples/collections. For such objects, the typechecks are skipped.

Cpp-driver performs typechecks during binding - we will follow the same approach. It is much convenient and cleaner to handle untyped objects this way.

## Implemented API functions
- `cass_collection_new_from_data_type` -> allows user to create a TYPED collection. We also adjusted the `cass_collection_new` logic - this function is responsible for creating an UNTYPED version of a collection.
- `cass_collection_data_type` -> returns a data type representing a collection. Since collections now hold info about the data types, this function was very trivial to implement.

## Unit tests
I implemented unit tests for simple and complex (udt, tuples, collections) types. Please make sure that all edge cases are covered.

## Pre-review checklist

<!--
    Make sure you took care of the issues on the list.
    Put 'x' into those boxes which apply.
    You can also create the PR now and click on all relevant checkboxes.
-->

- [x] I have split my patch into logically separate commits.
- [x] All commit messages clearly explain what they change and why.
- [x] PR description sums up the changes and reasons why they should be introduced.
- [x] I have implemented Rust unit tests for the features/changes introduced.
- ~[ ] I have enabled appropriate tests in `.github/workflows/build.yml` in `gtest_filter`.~
- ~[ ] I have enabled appropriate tests in `.github/workflows/cassandra.yml` in `gtest_filter`~

muzarski commented:
v1.1: Added a link to cpp-driver's UDT typecheck logic in a comment.

Lorak-mmk commented on scylla-rust-wrapper/src/value.rs:
```diff
@@ -89,7 +89,16 @@ impl CassCqlValue {
-            CassCqlValue::Text(_) => todo!(),
+            CassCqlValue::Text(_) => {
+                matches!(
+                    typ.get_value_type(),
+                    CassValueType::CASS_VALUE_TYPE_TEXT
+                        | CassValueType::CASS_VALUE_TYPE_VARCHAR
+                        | CassValueType::CASS_VALUE_TYPE_ASCII
+                        | CassValueType::CASS_VALUE_TYPE_BLOB
+                        | CassValueType::CASS_VALUE_TYPE_VARINT
+                )
+            }
```
cpp-driver also includes `CASS_VALUE_TYPE_CUSTOM`. I assume we are skipping support for this. Plase not this in the commit message here and in the commit about blobs.

muzarski replied:
Done

Lorak-mmk commented on scylla-rust-wrapper/src/cass_types.rs:
```diff
@@ -116,6 +116,26 @@ pub enum CassDataType {
+            CassDataType::Set { .. } => todo!(),
+            CassDataType::Map { .. } => todo!(),
+            CassDataType::Tuple(_) => todo!(),
+            CassDataType::Custom(name) => match other {
+                CassDataType::Custom(other_name) => name == other_name,
+                _ => false,
+            },
```
Here I see that there is some handling for custom types, so my previous comment may be incorrect.
Do we want to support them?

muzarski replied:
I was surprised that there is a CassDataType::Custom variant as well. If we don't plan on supporting it, we should remove it I think. AFAIR, custom types are not planned to be supported in the Rust driver.

muzarski replied:
I opened an issue: https://github.com/scylladb/cpp-rust-driver/issues/158

muzarski replied:
I'll remove mentions of custom types in a separate PR.

Lorak-mmk replied:
If we are not going to support Custom types then I see no point in introducing code related to it. Please replace this check with some panic.

muzarski replied:
Done

Lorak-mmk commented on scylla-rust-wrapper/src/cass_types.rs:
```diff
@@ -116,6 +116,26 @@ pub enum CassDataType {
+            CassDataType::Custom(name) => match other {
+                CassDataType::Custom(other_name) => name == other_name,
+                _ => false,
+            },
```
It will probably become clear in the next commit, but this one in isolation is not clear to me. Could you explain in comment or commit message (or both) why do we need `typecheck_equals`? Previous commits already implemented type checking, so it's not clear to me as a reviewer where this is going to be used.

muzarski replied:
This is used for complex types (udts, tuples, collections) in `value::is_type_compatible`. I'll include this information in the commit message.

Lorak-mmk commented on scylla-rust-wrapper/src/cass_types.rs:
```diff
@@ -112,6 +112,7 @@ pub enum CassDataType {
         val_type: Option<Arc<CassDataType>>,
         frozen: bool,
     },
+    // Empty vector stands for untyped tuple.
```
I know that this PR didn't introduce this, but I wonder if this is correct.
Is it possible to have an empty tuple in Scylla? If so, it could be a problem because this code won't differentiate between those 2 cases.

muzarski replied:
> Is it possible to have an empty tuple in Scylla?

It doesn't seem like it:
![image](https://github.com/user-attachments/assets/94c291cd-cb5a-442a-821b-e5ff5ef068dc)

I also checked creating an empty tuple type (`tuple<>`) using cqlsh. Scylla rejects such type.

Lorak-mmk commented on README.md:
Regarding commit `collection: include info about data type `: there is typo in the message.
`cass_collection_new_from_datat_type` should be `cass_collection_new_from_data_type`

Lorak-mmk commented on scylla-rust-wrapper/src/collection.rs:
```diff
@@ -32,8 +32,49 @@ pub struct CassCollection {
+                } => {
+                    // Only do the typecheck if both map types are present.
+                    if let (Some(k_typ), Some(v_typ)) = (k_typ, v_typ) {
+                        if index % 2 == 0 && !value::is_type_compatible(value, k_typ) {
+                            return CassError::CASS_ERROR_LIB_INVALID_VALUE_TYPE;
+                        }
+                        if !value::is_type_compatible(value, v_typ) {
+                            return CassError::CASS_ERROR_LIB_INVALID_VALUE_TYPE;
+                        }
```
This condition looks wrong. If `index % 2 == 0` and `value` is compatible with `k_typ` then we move to second if and require that it is also compatible with `v_typ`.

Didn't tests catch this? Please make sure they do.

muzarski commented:
v1.2: rebased on master

muzarski commented:
v2: addressed some review comments

What should we do with the custom types? @Lorak-mmk

Lorak-mmk commented:
I think we should either support them fully (which would require some changes in Rust Driver, probably not big ones) or not support them at all - remove them from enums, return errors as early as possible when encountering them somewhere. WDYT @wprzytula ?

muzarski commented:
Rebased on master

muzarski commented:
Rebased on master (decimal)

muzarski commented:
v1.3: added mention about not supporting custom types in Text and Blob commits.

wprzytula commented on scylla-rust-wrapper/src/collection.rs:
```diff
@@ -57,18 +75,61 @@ pub unsafe extern "C" fn cass_collection_new(
 ) -> *mut CassCollection {
     let capacity = match collection_type {
         // Maps consist of a key and a value, so twice
-        // the number of CqlValue will be stored.
+        // the number of CassCqlValue will be stored.
```
I think the previous formulation was correct. If `CqlValue::Map` has capacity `n`, then `CassCollection` functioning as a Map has capacity `2n`.

muzarski replied:
I only changed `CqlValue` -> `CassCqlValue`.

wprzytula commented on scylla-rust-wrapper/src/collection.rs:
```diff
@@ -57,18 +75,61 @@ pub unsafe extern "C" fn cass_collection_new(
     }))
 }
 
+#[no_mangle]
+unsafe extern "C" fn cass_collection_data_type(
+    collection: *const CassCollection,
+) -> *const CassDataType {
+    let collection_ref = ptr_to_ref(collection);
+
+    match &collection_ref.data_type {
+        Some(dt) => Arc::as_ptr(dt),
+        None => match collection_ref.collection_type {
+            CassCollectionType::CASS_COLLECTION_TYPE_LIST => &UNTYPED_LIST_TYPE,
+            CassCollectionType::CASS_COLLECTION_TYPE_SET => &UNTYPED_SET_TYPE,
+            CassCollectionType::CASS_COLLECTION_TYPE_MAP => &UNTYPED_MAP_TYPE,
+            // CassCollectionType is a C enum. Can user provide a value out of range?
+            _ => unreachable!(),
```
User can provide a value out of range, but then it's going to be their fault. That's how C contract on UB works...

muzarski replied:
I think we should panic here then (with some insightful error message). I wouldn't silently return a null pointer.

wprzytula commented on scylla-rust-wrapper/src/collection.rs:
```diff
@@ -1,14 +1,32 @@
 
 include!(concat!(env!("OUT_DIR"), "/cppdriver_data_collection.rs"));
 
+/// These constants help us to save an allocation in case user calls `cass_collection_new` (untyped collection).
+static UNTYPED_LIST_TYPE: CassDataType = CassDataType::List {
+    typ: None,
+    frozen: false,
+};
+static UNTYPED_SET_TYPE: CassDataType = CassDataType::Set {
+    typ: None,
+    frozen: false,
+};
+static UNTYPED_MAP_TYPE: CassDataType = CassDataType::Map {
+    key_type: None,
+    val_type: None,
+    frozen: false,
+};
```
Great idea! Love it!

wprzytula commented on scylla-rust-wrapper/src/collection.rs:
```diff
@@ -1,14 +1,32 @@
 
 include!(concat!(env!("OUT_DIR"), "/cppdriver_data_collection.rs"));
 
+/// These constants help us to save an allocation in case user calls `cass_collection_new` (untyped collection).
+static UNTYPED_LIST_TYPE: CassDataType = CassDataType::List {
+    typ: None,
+    frozen: false,
+};
+static UNTYPED_SET_TYPE: CassDataType = CassDataType::Set {
+    typ: None,
+    frozen: false,
+};
```
Nit: this probably shouldn't be a docstring, because it would only document the first of the constants.

wprzytula commented on scylla-rust-wrapper/src/collection.rs:
```diff
@@ -57,18 +75,61 @@ pub unsafe extern "C" fn cass_collection_new(
+    }))
+}
+
+#[no_mangle]
+unsafe extern "C" fn cass_collection_new_from_data_type(
+    data_type: *const CassDataType,
+    item_count: size_t,
+) -> *mut CassCollection {
+    let data_type = clone_arced(data_type);
+    let (capacity, collection_type) = match data_type.as_ref() {
+        CassDataType::List { .. } => (item_count, CassCollectionType::CASS_COLLECTION_TYPE_LIST),
+        CassDataType::Set { .. } => (item_count, CassCollectionType::CASS_COLLECTION_TYPE_SET),
+        // Maps consist of a key and a value, so twice
+        // the number of CassCqlValue will be stored.
+        CassDataType::Map { .. } => (item_count * 2, CassCollectionType::CASS_COLLECTION_TYPE_MAP),
+        _ => return std::ptr::null_mut(),
```
Does cpp driver silently return a NULL pointer on bad data type passed? If so, that's sad...

wprzytula commented on scylla-rust-wrapper/src/collection.rs:
```diff
@@ -32,8 +32,49 @@ pub struct CassCollection {
+        if let Some(data_type) = &self.data_type {
+            match data_type.as_ref() {
+                CassDataType::List { typ: subtype, .. }
+                | CassDataType::Set { typ: subtype, .. } => match subtype {
+                    Some(subtype) => {
+                        if !value::is_type_compatible(value, subtype) {
+                            return CassError::CASS_ERROR_LIB_INVALID_VALUE_TYPE;
+                        }
+                    }
+                    None => {}
+                },
```
`if let` ?

muzarski commented:
Rebased on master

muzarski commented on scylla-rust-wrapper/src/tuple.rs:
```diff
@@ -50,7 +50,10 @@ impl CassTuple {
 
 impl From<&CassTuple> for CassCqlValue {
     fn from(tuple: &CassTuple) -> Self {
-        CassCqlValue::Tuple(tuple.items.clone())
+        CassCqlValue::Tuple {
+            data_type: tuple.data_type.clone(),
+            fields: tuple.items.clone(),
+        }
     }
 }
```
Sure, I can change it to `to_cass_cql_value` method. But I'd do it in a follow-up PR since such `From` implementation appears in multiple places AFAIR.

Lorak-mmk replied:
And of course this implementation is my fault :( https://github.com/scylladb/cpp-rust-driver/commit/db1a9a7f610f070ad60a3175cec4a9b11f579542

muzarski replied:
I opened an issue to remember about it: https://github.com/scylladb/cpp-rust-driver/issues/177

muzarski commented on scylla-rust-wrapper/src/cass_types.rs:
```diff
@@ -181,7 +183,33 @@ impl CassDataType {
+                    ((Some(k), Some(v)), (Some(k_other), Some(v_other))) => {
+                        k.typecheck_equals(k_other) && v.typecheck_equals(v_other)
+                    }
+                    _ => false,
```
Introduced `MapDataType` enum (last commit) which represents valid states of map data type.

muzarski commented:
v3: addressed review comments.

Lorak-mmk reviewed (changes requested):
One small comment, otherwise looks good to me

wprzytula commented on scylla-rust-wrapper/src/collection.rs:
```diff
@@ -58,11 +60,15 @@ impl TryFrom<&CassCollection> for CassCqlValue {
                 }
 
-                Ok(CassCqlValue::Map(grouped_items))
-            }
-            CassCollectionType::CASS_COLLECTION_TYPE_SET => {
-                Ok(CassCqlValue::Set(collection.items.clone()))
+                Ok(CassCqlValue::Map {
+                    data_type,
+                    values: grouped_items,
+                })
             }
```
Oh, you've fixed a bug here, correct? Map collections were made `CassCqlValue::Set` ...

muzarski replied:
Noo, I think it's just a confusing diff generated by git.

wprzytula commented on scylla-rust-wrapper/src/value.rs:
```diff
@@ -130,9 +139,27 @@ impl CassCqlValue {
-            CassCqlValue::List(_) => todo!(),
-            CassCqlValue::Map(_) => todo!(),
-            CassCqlValue::Set(_) => todo!(),
+            CassCqlValue::List { data_type, .. } => {
+                if let Some(dt) = data_type {
+                    return dt.typecheck_equals(typ);
+                }
+                // Untyped list.
+                typ.get_value_type() == CassValueType::CASS_VALUE_TYPE_LIST
+            }
+            CassCqlValue::Map { data_type, .. } => {
+                if let Some(dt) = data_type {
+                    return dt.typecheck_equals(typ);
+                }
+                // Untyped map.
+                typ.get_value_type() == CassValueType::CASS_VALUE_TYPE_MAP
+            }
+            CassCqlValue::Set { data_type, .. } => {
+                if let Some(dt) = data_type {
+                    return dt.typecheck_equals(typ);
+                }
+                // Untyped set.
+                typ.get_value_type() == CassValueType::CASS_VALUE_TYPE_SET
```
I'd strongly prefer to see this without such imperative `return`s, because such `return`s usually signify an abnormal return (an error, or a shortcut) and here that's no the case. Let's use functional patterns.

wprzytula replied:
```suggestion
            CassCqlValue::List { data_type, .. } => {
                if let Some(dt) = data_type {
                    dt.typecheck_equals(typ)
                } else {
                    // Untyped list.
                    typ.get_value_type() == CassValueType::CASS_VALUE_TYPE_LIST
                }
            }
            CassCqlValue::Map { data_type, .. } => {
                if let Some(dt) = data_type {
                    dt.typecheck_equals(typ)
                } else {
                    // Untyped map.
                    typ.get_value_type() == CassValueType::CASS_VALUE_TYPE_MAP
                }
            }
            CassCqlValue::Set { data_type, .. } => {
                if let Some(dt) = data_type {
                    dt.typecheck_equals(typ)
                } else {
                    // Untyped set.
                    typ.get_value_type() == CassValueType::CASS_VALUE_TYPE_SET
                }
```

wprzytula commented on scylla-rust-wrapper/src/value.rs:
```diff
@@ -725,5 +756,110 @@ mod tests {
+                        data_type_float_int_map.clone(),
+                    ],
+                },
+                // Typed list.
+                TestCase {
+                    value: CassCqlValue::List {
+                        data_type: Some(data_type_float_list.clone()),
+                        values: vec![],
+                    },
+                    compatible_types: vec![
+                        data_type_float_list.clone(),
+                        data_type_untyped_list.clone(),
+                    ],
+                    incompatible_types: vec![
+                        data_type_float.clone(),
+                        data_type_int.clone(),
+                        data_type_bool.clone(),
+                        data_type_untyped_set.clone(),
+                        data_type_float_set.clone(),
+                        data_type_untyped_map.clone(),
+                        data_type_typed_key_float_map.clone(),
+                        data_type_float_int_map.clone(),
+                    ],
+                },
```
Let's add an incompatible list type, e.g. `data_type_int_list`.

wprzytula commented on scylla-rust-wrapper/src/value.rs:
```diff
@@ -725,5 +756,110 @@ mod tests {
+                        data_type_float_int_map.clone(),
+                    ],
+                },
+                // Only key-typed map.
+                TestCase {
+                    value: CassCqlValue::Map {
+                        data_type: Some(data_type_typed_key_float_map.clone()),
+                        values: vec![],
+                    },
+                    compatible_types: vec![
+                        data_type_typed_key_float_map.clone(),
+                        data_type_untyped_map.clone(),
+                    ],
+                    incompatible_types: vec![
+                        data_type_float.clone(),
+                        data_type_int.clone(),
+                        data_type_bool.clone(),
+                        data_type_float_list.clone(),
+                        data_type_untyped_list.clone(),
+                        data_type_untyped_set.clone(),
+                        data_type_float_set.clone(),
+                        data_type_float_int_map.clone(),
+                    ],
+                },
+            ];
+
+            run_test_cases(test_cases)
+        }
```
Tests for untyped map and fully typed map are missing. Similarly, tests for set are missing.

muzarski replied:
Typed maps and sets are included in `[in]compatible_types` for some other types. But I'll introduce separate test cases for them as well.

wprzytula commented on scylla-rust-wrapper/src/collection.rs:
```diff
@@ -32,8 +32,48 @@ pub struct CassCollection {
+                            return CassError::CASS_ERROR_LIB_INVALID_VALUE_TYPE;
+                        }
+                    }
+                }
```
So you say that we shouldn't check key type compatibility in case of a key-only-typed map?

muzarski replied:
I know that this is confusing (since we allow users to define such map types), but this is the cpp-driver's logic:
https://github.com/scylladb/cpp-driver/blob/master/src/collection.hpp#L100-L110

muzarski replied:
The `if (data_type_->types().size() == 2` part implies that both types need to be present to bind value to map collection.

wprzytula replied:
Do we prefer to be mimic cpp-driver as tight as possible in this case, or maybe let's be more eager and do the type check of the key, if present, to save the user from errors?

muzarski replied:
To be honest, I think this is a bug in a cpp-driver. We should not mimic this behaviour

wprzytula commented on scylla-rust-wrapper/src/collection.rs:
```diff
@@ -163,3 +203,172 @@ make_binders!(decimal, cass_collection_append_decimal);
+    #[test]
+    fn test_typecheck_on_append_to_collection() {
+        unsafe {
+            // untyped map
+            {
+                let untyped_map =
+                    cass_collection_new(CassCollectionType::CASS_COLLECTION_TYPE_MAP, 2);
+                assert_cass_error_eq!(
+                    cass_collection_append_bool(untyped_map, false as cass_bool_t),
+                    CassError::CASS_OK
+                );
+                assert_cass_error_eq!(
+                    cass_collection_append_int16(untyped_map, 42),
+                    CassError::CASS_OK
+                );
+                assert_cass_error_eq!(
+                    cass_collection_append_double(untyped_map, 42.42),
+                    CassError::CASS_OK
+                );
+                assert_cass_error_eq!(
+                    cass_collection_append_float(untyped_map, 42.42),
+                    CassError::CASS_OK
+                );
+                cass_collection_free(untyped_map);
+            }
+
+            // typed map
+            {
```
What about a half-typed (key-only-typed) map?

wprzytula commented on scylla-rust-wrapper/src/statement.rs:
```diff
@@ -44,13 +44,30 @@ pub struct CassStatement {
         }
+
+        self.bound_values[index] = Set(value);
+        CassError::CASS_OK
```
I dislike that we perform indexing, which may panic. Let's try to refactor this so that we match use `get()` method on slice and pattern match on its result. The pattern matching should involve both `self.bound_values` and `p.variable_col_data_types`, because we index both.

muzarski replied:
I refactored the code

wprzytula commented on scylla-rust-wrapper/src/cass_types.rs:
```diff
@@ -338,10 +337,19 @@ impl CassDataType {
+                typ: MapDataType::Untyped,
+                ..
+            } => None,
+            CassDataType::Map {
+                typ: MapDataType::Key(k),
+                ..
+            } if index == 0 => Some(k),
```
To save the readers from reasoning about remaining, non-exhausted patterns, let's do it this way:

wprzytula replied:
```suggestion
            CassDataType::Map {
                typ: MapDataType::Key(k),
                ..
            } => (index == 0).then_some(k),
```

wprzytula commented on scylla-rust-wrapper/src/collection.rs:
```diff
@@ -47,19 +46,17 @@ impl CassCollection {
+                        return CassError::CASS_ERROR_LIB_INVALID_VALUE_TYPE;
                     }
                 }
                 _ => return CassError::CASS_ERROR_LIB_INVALID_VALUE_TYPE,
```
Bug spotted! Before this commit, untyped and half-typed map would be accepted, and now they are rejected (they don't match `MapDataType::KeyAndValue`).

muzarski replied:
They wouldn't be accepted - previous version included a check for this:
```rust
// Only do the typecheck if both map types are present.
if let (Some(k_typ), Some(v_typ)) = (k_typ, v_typ) {
```

muzarski replied:
Sorry, I misunderstood.

muzarski replied:
I'll fix it and introduce additional test cases. Good catch!

muzarski commented:
Rebased on master

muzarski commented on scylla-rust-wrapper/src/user_type.rs:
```diff
@@ -20,7 +19,7 @@ impl CassUserType {
             return CassError::CASS_ERROR_LIB_INDEX_OUT_OF_BOUNDS;
         }
-        if !is_compatible_type(&self.data_type.get_udt_type().field_types[index].1, &value) {
+        if !value::is_type_compatible(&value, &self.data_type.get_udt_type().field_types[index].1) {
```
Since I moved it to `value` module, I thought that it makes more sense to have a value as first parameter.

muzarski commented on scylla-rust-wrapper/src/value.rs:
```diff
@@ -55,8 +55,7 @@ pub enum CassCqlValue {
     Set(Vec<CassCqlValue>),
     UserDefinedType {
-        keyspace: String,
-        type_name: String,
+        data_type: Arc<CassDataType>,
```
That's true - `UDTDataType` would be sufficient. However, this way it's much easier to call `CassDataType::typecheck_equals`. In addition, we would need to clone whole `UDTDataType` since it's owned by a shared reference to `CassDataType` (see `From<&CassUserType> for CassCqlValue` implementation). As of now, we only need to clone an Arc<CassDataType>, which is much cheaper operation.

muzarski commented:
v4: addressed @wprzytula comments. Thanks for spotting the bug! I introduced a lot of additional test cases which consider more combinations during typechecks
